test/lib/jdk/test/lib/JDWP.java

```diff
@@ -49,4 +49,17 @@ public static ListenAddress parseListenAddress(String debuggeeOutput) {
         return null;
     }
 
+    /**
+     * Parses debuggee output to get listening transport and address, printed by `launch=echo`.
+     * Returns null if the string specified does not contain required info.
+     */
+    public static ListenAddress parseLaunchEchoListenAddress(String debuggeeOutput) {
+        String[] parts = debuggeeOutput.split(" ");
+        if (parts.length != 2) {
+            return null;
+        }
```

Comment:
This is a bit presumptuous in that it assumes that the first 2 token line of output is the line we are looking for. The reason `parseListenAddress()` tries to match a specific pattern is because sometimes tests are run in a way that causes the debuggee to produce some extra output (like logging output) and this output needs to be skipped.  At the very least you should be checking for a pattern that includes "dt_socket".

BTW, I tried using `printf` instead of `echo` so the debug agent could be made to generate a "Listening for transport..." line, in which case this new `parseLaunchEchoListenAddress()` would not be needed, but I could not get the argument processing right.  Something about exec() and printf just doesn't work right.

Reply:
Making too many not enough assumptions. The other method didn't make assumptions on the format of address and connection type, so I thought that I shouldn't too. I could create a temporary shell file that mimicks the other output though.

Reply:
`parseListenAddress()` looks for a fairly specific format:

`listenRegexp = Pattern.compile("Listening for transport \\b(.+)\\b at address: \\b(.+)\\b");`

However, it does not assume the transport type, so maybe doing that here is not such a good idea. I don't like adding shell scripts to tests. We've been converting tests from shell scripts for years now because they tend to be problematic.

How about using something like launch="echo Listen Args: ". then you have a more specific string you can pattern match on:

`Listen Args: dt_socket 42305`

Reply:
Maybe you can use call `enableOnThrow()` as a trigger to use a specific echo command whose pattern is later detected. I'm not so sure `parseLaunchEchoListenAddress()` needs to be in JDWP since it will only be used by the Debuggee class. That spares the JDWP class from needing to know about this pattern.

Reply:
Good to know that special shell scripts are to be avoided. I agree that your proposal is better than just the simple ’echo’.

I'll implement it tomorrow morning.